### ISSUE TITLE
Registering a non-standard identification function

### ISSUE THREAD

**Issue by zibrov-zlobin** [labrad/pylabrad#126]
We have an Amatek Lock-in. Instead of the standard "*IDN?" command the identification command is : "ID". 
The "gpib_device_manager.py" has a method `register_ident_function()` that has the setting defined as `setting = ['s','w']`.
As I understand, I should be able to register a new function by connecting to the server and invoking this method. 
```python
import labrad
cxn  = labrad.connect()
cxn.gpib_device_manager.register_ident_function("AMATEK Lock-in Server, GPIB0::12::INST")=[]
```
where "AMATEK Lock-in Server" is the name of the device server running and connected to the LabRAD manager.
but where do I provide the new call signatures? And why isnt' the float ('w') required in the setting? 
is this  function registration supposed to be added as a client in the device server? 

I guess I don't understand the `@setting` decorator and the datatypes. Any method that has a "setting=" is expecting a list? e.g.
```python
@setting(101,'FuncName', setting=['s','w'])
def method1(self,c,setting):
    pass
```
does it mean that to invoke the method I would do: `method1(c,["bla","45.0"])` ?

**Comment by DanielSank** [comment]
On the setting decorator issue:

If you do this
```python
@setting(101, "Name", arg=['s','w'])
def name(self, c, arg):
    ...
```
you are telling pylabrad that you have a remotely accessible setting called `Name` which takes a single argument called `arg` which can be either a string (`s`) or an unsigned int (`w`).

**Comment by maffoo** [comment]
The idea here is that you need to write a new server to manage the gpib communication with your amatek lock-in devices. Your server registers with the gpib_device_manager that it can handle certain devices, but the gpib_device_manager itself handles identifying all the available gpib devices, and then notifying registered servers when it has found them. For non-standard cases, your server can implement its own identification logic as a setting that the gpib_device_manager will call when it finds devices devices that do not respond to `*IDN?` queries, for example. The best way to make this work is to have your server subclass `labrad.gpib.GPIBManagedServer`. The documentation for implementing such a server, and for how the interaction with the gpib_device_manager works, needs to be improved. Also, we should probably pull the gpib stuff out of labrad and into its own labrad-gpib package along with the gpib_device_manager in the servers repo, because the gpib stuff is not really usable without the device manager and vice versa.

**Comment by zibrov-zlobin** [comment]
Thanks for the replies. I guess I dont' quite understand how to access the settings. 
For example, I am trying to measure the voltage with the Agilent 34401A DMM.

The `AgilentDMMServer` has a method:
```python
@setting(10, AC='b{AC}', returns='v[V]')
def voltage(self,c,AC=False):
```
I am trying to access it by, but I don't understand what arguments (in what form) is the server accepting; I keep getting "could not flatten errors" or "key unkown":

```python
import labrad
cxn = labrad.connect()
server = cxn.agilent_34401a_dmm
c = cxn.context()
server.voltage(c, True )
```
I am sorry to bother with this, it seems  to be basic  but I can't figure it out. Could anyone give an example of a client request? 
thanks

**Comment by DanielSank** [comment]
That first argument `c` in the `voltage` function is passed in automatically by pylabrad. You should not pass it yourself from the interactive shell. Instead, do this
```python
import labrad
cxn = labrad.connect()
server = cxn.agilent_34401a_dmm
result = server.voltage(True)
```

**Comment by DanielSank** [comment]
NB: Issues like this should be closed only once OP understands what to do and either he/she or someone else has condensed the relevant information into a wiki article.

**Comment by DanielSank** [comment]
@zibrov-zlobin Don't worry for now about using non-default contexts. You will probably never need the features they provide.

**Comment by zibrov-zlobin** [comment]
@DanielSank Thanks! 
I wasn' t sure if it was me making incorrect requests to the server or the code. 
The agilent_34401a_dmm server had an issue with flattening the response to the set unit.

I've added 
```python
import labrad.units as units
```
and replaced all the `returnValue(float(ans))` to `returnValue(float(ans)*units.ApropriateUnit)`

**Comment by zibrov-zlobin** [comment]
@maffoo 
Thanks! It works!

will try to write up a wiki, do I understand it right:
if my class has an attribute `self.deviceIdentFunc` which points to an identification method then the server registers the name of this method in the `gpib_device_manager`. If the regular *IDN? identification results in UNKNOWN, the device manager tries with the method specified? 

Then I don't understand what does the docstring mean by:

```python
        This setting must accept either of the following:
        
            s, s, s: server, address, *IDN? response
            s, s:    server, address
```
i guess (in retrospec) this means these are the requirements for the identification method header, not the argument passed to the `register_ident_function` ?

**Comment by maffoo** [comment]
@zibrov-zlobin, yeah, that doc means that your server with ident setting should look like this:

```python
deviceIdentFunc = 'my_ident_func'

@setting(1000, server='s', address='s', idn_response='s')
def my_ident_func(self, c, server, address, idn_response=None):
    ...
```

By adding a default value of the `idn_response` argument, we tell labrad that it is optional, so this function will accept `(s, s)` or `(s, s, s)`, as desired.
